### ISSUE TITLE
[miio] OutOfMemory errors due to threads being continuously created

### ISSUE THREAD

**Issue by kaikreuzer** [openhab/openhab-addons#8091]
I am seeing regular OOM errors on my instance and just had a look at the thread dump. What I am seeing is that the MI IO binding seems to be the culprit as it creates hundreds of threads:
```
"Mi IO MessageSenderThread" #40012 daemon prio=5 os_prio=31 tid=0x00007fb1ebca2000 nid=0x4aa67 waiting on condition [0x000070001ff01000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39931 daemon prio=5 os_prio=31 tid=0x00007fb1ec60e800 nid=0x44bd3 waiting on condition [0x000070002329a000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39843 daemon prio=5 os_prio=31 tid=0x00007fb1ec428000 nid=0x4b283 sleeping[0x000070001fdfe000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39711 daemon prio=5 os_prio=31 tid=0x00007fb1f3335000 nid=0x4e697 waiting on condition [0x000070002236d000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39616 daemon prio=5 os_prio=31 tid=0x00007fb1edc17800 nid=0x4abc7 waiting on condition [0x0000700023197000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39543 daemon prio=5 os_prio=31 tid=0x00007fb1ee94b800 nid=0x427db waiting on condition [0x0000700022779000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39465 daemon prio=5 os_prio=31 tid=0x00007fb1eeb4a800 nid=0x2eacb waiting on condition [0x0000700020b25000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39321 daemon prio=5 os_prio=31 tid=0x00007fb1f3668800 nid=0x4d58f sleeping[0x0000700021c58000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39212 daemon prio=5 os_prio=31 tid=0x00007fb1ef47b000 nid=0x3e8df waiting on condition [0x0000700020719000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39129 daemon prio=5 os_prio=31 tid=0x00007fb1ec0c3800 nid=0x4037f sleeping[0x0000700020004000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #39032 daemon prio=5 os_prio=31 tid=0x00007fb1f33fe800 nid=0x4787b sleeping[0x000070002020a000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #38909 daemon prio=5 os_prio=31 tid=0x00007fb1ef6a5800 nid=0x49d43 waiting on condition [0x0000700020e2e000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #38809 daemon prio=5 os_prio=31 tid=0x00007fb1ed889800 nid=0x4da17 waiting on condition [0x0000700021a52000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #38725 daemon prio=5 os_prio=31 tid=0x00007fb1ef260000 nid=0x3fa0b sleeping[0x000070002123a000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #38626 daemon prio=5 os_prio=31 tid=0x00007fb1f196c800 nid=0x4336b waiting on condition [0x000070001d380000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #38507 daemon prio=5 os_prio=31 tid=0x00007fb1ed9c7000 nid=0x49947 waiting on condition [0x000070002226a000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #38396 daemon prio=5 os_prio=31 tid=0x00007fb1f026b000 nid=0x4a4bb waiting on condition [0x000070001c85f000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)

"Mi IO MessageSenderThread" #38328 daemon prio=5 os_prio=31 tid=0x00007fb1f04d0800 nid=0x44973 waiting on condition [0x000070002184c000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openhab.binding.miio.internal.transport.MiIoAsyncCommunication$MessageSenderThread.run(MiIoAsyncCommunication.java:225)
...
```

**Comment by lolodomo** [comment]
I fixed one memory leak in this binding with #7607 
Either this fix was overwritten or there are other leaks in the binding.

**Comment by kaikreuzer** [comment]
Thanks, the report #7583 indeed looks pretty much the same.
The bad news is that it does not seem to be fixed by #7607.
My system is running a build from May 27, so the fix should be included there already.

**Comment by marcelrv** [comment]
@kaikreuzer do you still experience the issue or did #8189 fix it?
